### DISCUSSION TITLE
Iterator support

### DISCUSSION THREAD

**Discussion by jethrogb** [diesel-rs/diesel#2707] [Ideas]
I'd like to issue select queries that return all data lazily in the form of an iterator. I want to issue queries for a million rows or so and I don't want to allocate all the memory up front.

This seems to have been the way things worked, but it was changed in #132/#142. I don't fully understand the rationale for the change. I also don't care about associations for this particular query.

**Comment by vandenoever** [comment]
I'd very much appreciate the convenience of ORM without the memory allocation overhead. An iterator that returns shallow items that use `&str` instead of `String` would be good in low-memory use-cases.

**Comment by deankarn** [comment]
I was very much enjoying learning and experimenting with diesel until I found that it returned a `Vec` rather than an `Iterator`.

This limitation makes this library unusable in a lot of high performance, low memory and future asyncio situations.

Is there a workaround or method I'm missing to accomplish this? It would also be great if someone could explain the rationale behind moving away from iterators, perhaps there's a good reason I'm missing?

**Comment by lnicola** [comment]
Note that it's often considered a good idea to buffer the results. Reading row by row, often tends to keep the query running for a longer time, possibly decreasing concurrency on the server. Realistically, the cost of the allocation is tiny compared to the network and query latency.

Of course, buffering works better with small result sets or pagination. But if you're reading a million rows at once, there might be better approaches.

**Comment by deankarn** [comment]
@lnicola thanks for the quick response 😄

I agree in many situations such a large one time query is often avoidable. I was hoping though that the option would still be available for the user in the cases where it's unavoidable.

Thanks again!

**Comment by richard-uk1** [comment]
I actually wrote some code to do this today. I'll copy it verbatim below. It uses an internal buffer and offset/limit. It would be good if I could pass my buffer to a method rather than have it allocate a new Vec every time, but I don't think this is possible currently.

```rust
//! A little abstraction to fetch database records in batches.
//!
//! TODO this currently will fail if there are concurrent connections. It should use some sort of
//! lock.
use diesel::{
    dsl::{Limit, Offset},
    prelude::*,
    query_dsl::{
        methods::{LimitDsl, OffsetDsl},
        LoadQuery,
    },
    result::Error,
};
use std::{collections::VecDeque, convert::TryInto, marker::PhantomData};

/// Get an object that implements the iterator interface.
pub fn stream_results<'conn, Record, Query, Conn>(
    query: Query,
    conn: &'conn Conn,
    buffer_size_hint: usize,
) -> impl Iterator<Item = Result<Record, Error>> + 'conn
where
    Record: 'conn,
    Query: OffsetDsl + Clone + 'conn,
    Offset<Query>: LimitDsl,
    Limit<Offset<Query>>: LoadQuery<Conn, Record>,
{
    RecordIter {
        conn,
        query,
        cursor: 0,
        buffer: VecDeque::with_capacity(buffer_size_hint),
        record_type: PhantomData,
    }
}

struct RecordIter<'conn, Record, Query, Conn> {
    conn: &'conn Conn,
    query: Query,
    /// The index of the next record to fetch from the server
    cursor: usize,
    buffer: VecDeque<Record>,
    record_type: PhantomData<Record>,
}

impl<'conn, Record, Query, Conn> Iterator for RecordIter<'conn, Record, Query, Conn>
where
    Query: OffsetDsl + Clone,
    Offset<Query>: LimitDsl,
    Limit<Offset<Query>>: LoadQuery<Conn, Record>,
{
    type Item = Result<Record, Error>;

    fn next(&mut self) -> Option<Self::Item> {
        // if the buffer isn't empty just return an element
        match self.buffer.pop_front() {
            Some(v) => return Some(Ok(v)),
            None => (), // fall through
        };
        // fill the buffer
        let fetch_amt = self.buffer.capacity();
        let query = self
            .query
            .clone()
            .offset(self.cursor.try_into().unwrap())
            .limit(fetch_amt.try_into().unwrap());
        self.cursor += fetch_amt;
        let results: Vec<Record> = match query.load(&self.conn) {
            Ok(recs) => recs,
            Err(e) => return Some(Err(e)),
        };
        for result in results {
            self.buffer.push_back(result);
        }
        // return the first record, or None if there are no more records fetched.
        self.buffer.pop_front().map(Ok)
    }
}
```

**Reply by richard-uk1** [reply]
This isn't the best way to do iterators over results. Using `OFFSET` requires the db iterating through all the results until the offset of interest, which is very slow. Database engines provide cursor objects to manage fetching rows in a streaming manner.

**Comment by chris13524** [comment]
I'm also interested in why we aren't using Iterators here? If you want to process all records you might not be able to pull that into memory. IMO it would make sense to expose an Iterator where underneath it runs `offset()` and `limit()` to pull records in batches on-demand.

**Comment by weiznich** [comment]
@chris13524 We are not interested in supporting iterators build on top of `offset` and `limit` as this hide a lot of complexity about the query from the user. It would for example not be possible to write queries with their own offset or limit clause, as those would get overwritten by the iterator (or the iterator needs complicated special cases for those)
Additionally it is easily possible to build something like that on top of diesel, so I do not see any need to have this in diesel itself.

**Comment by mjanda** [comment]
You could probably use cursors to implement this.

https://www.postgresql.org/docs/11/plpgsql-cursors.html

But I guess it would be quite complex change and it might not be possible for all databases.

Some crates implement prefetching N rows at a time (rust-oracle for example).

**Comment by weiznich** [comment]
@mjanda See the two issues/PR's linked in OP. The original implementation provided a method that returned a iterator based on cursors. There are reasons listed there to switch that to return a `Vec<T>` instead. I'm not sure if all of the reasoning in there is relevant anymore as the API changed quite a bit since then. Internally all of our currently supported backends are using cursors.

That said at least for me there are currently more important things to work on than this, as I think it is not useful for most cases. For the general case you end up collecting the result somewhere anywhere.

**Comment by rbtcollins** [comment]
AIUI the key thing was that associations at the time had two related issues (detailed in #132):
- they didn't stream, instead reading the entire result to find associated keys up front
- they could fail but the iterated result type was not a fallible type, which meant things had to panic.

I think if someone wanted to work on this, verifying if those things are still the case would be a first step; followed by figuring out other ways to address those issues; once addressed, changing from Vec to Iterator should be relatively straight forward.

**Comment by weiznich** [comment]
So I've moved this into the discussion section as this is a feature request. The whole point of moving this now is that I spend some time looking into this issue again and I noticed the following things:
 
All our supported backends internally already use a cursor based API, so exposing would be possible in principle.
That written I believe it's practically impossible to implement that in a sound way with the current design of the `Connection` trait. Let me explain that while using the [`Connection::load`](https://github.com/diesel-rs/diesel/blob/3f3ce57ca1a89274ecf915b70826234c45bd261e/diesel/src/connection/mod.rs#L183) method as example. This method is used to load all data from the database and deserialize the result into some compatible type. First let's look at the mentioned method:
```rust
fn load<T, U>(&self, source: T) -> QueryResult<Vec<U>>
    where
        Self: Sized,
        T: AsQuery,
        T::Query: QueryFragment<Self::Backend> + QueryId,
        U: FromSqlRow<T::SqlType, Self::Backend>,
        Self::Backend: QueryMetadata<T::SqlType>
```

It does currently return a `Result<Vec<U>>` where `U` is our deserialized type. If we now want to return some iterator like API this opens the question how that return type would look like. I see two possibilities here: Either return a type that implements `Iterator<Item = Result<U>>` or a type that implements `Iterator<Item = Result<impl Row<Self::Backend>>`, where for the second case `impl Row<Self::Backend>` is some opaque type that exposes only the `Row` trait. Both imply that the iterator needs to have some sort of access to the underlying cursor, which is created from a prepared statement, which may be cached as part of the connection implementation. Now this raises the following issue. Let's assume the following rust code using an hypothetical streaming API:

```rust
let iter_1 = users::table.load_as_iter(&conn)?;
let iter_2 = users::table.load_as_iter(&conn)?;

let user = iter_1.collect();
```

Both queries will be bound to the same prepared statement internally, so this will expose the result of the second query instead of the first one. As consequence this means a iterator API needs to couple the lifetime of the returned iterator to unique access to the connection itself, which in rust can easily be expressed by a mutable borrow. Now the Problem: All our connection methods and corresponding `RunQueryDsl` methods use a shared reference, so this would need to change to allow such an API to exist. 

The question now is do we want to change the `Connection` trait (and everything build on-top) to require a mutable reference instead of a shared one? On the one hand this would unblock support for such "advanced" API's and would probably also be helpful for the async connection implementation. On the other hand that would mean a quite large change for all existing diesel code.
I should probably add that in practice that does not mean much as it is currently impossible to use connections from multiple threads simultaneously and there is no way to return a borrowed result from a connection, so each access is already unique, which means the change would mostly be that users needs to systematically replace all `&DbConnection` with `&mut DbConnection` in their code bases.


We need to decide to change that before releasing 2.0 as at least I do not plan to do another major release any time soon.

cc @diesel-rs/reviewers @diesel-rs/contributors

**Reply by Ten0** [reply]
> Either return a type that implements `Iterator<Item = Result<U>>` or a type that implements `Iterator<Item = Result<impl Row<Self::Backend>>`

It looks like `Iterator<Item = Result<U>>` would cover the the general case very well, be uniform with `load` and not be very restrictive in terms of usage since one could probably implement `FromSqlRow` for their `U` type.

It looks like returning an iterator that yields `impl Row<Self::Backend>` may be restrictive with regards to what row may be: currently it may have a lifetime tied to a variable used in the `load` function, that represents "here are some of the results I got from the database":
https://github.com/diesel-rs/diesel/blob/42ad5b3d9f78a9307b1c704ac4d3a87363bb4214/diesel/src/pg/connection/mod.rs#L79-L84
This could be extended to having a lifetime tied to somthing that's stored in the new iterator structure if it were to yield `Result<U>`, since it would anyway be dropped before the call to `next()` returns.
That's however not possible anymore if we have to be able to generate an iterator that returns it.

So overall I'd clearly favor `Iterator<Item = Result<U>>`.

**Reply by weiznich** [reply]
And what about the "large" change requiring a `&mut Connection` everywhere?

**Reply by lnicola** [reply]
That could be avoided for now using run-time checks, right? It's less compile-time checked, but a reasonable compromise nonetheless.

**Reply by pksunkara** [reply]
If this is the direction we want to take diesel in the future, it is better if we implement it now. Regarding the large changes of `&mut Connection`, maybe rust-analyzer has some refactoring assists?

**Reply by Ten0** [reply]
> And what about the "large" change requiring a `&mut Connection` everywhere?

I accidentally lost quite some time of message-writing by restarting my browser so I decided this time I'd submit it in two times as I was rewriting it ^^

> which is created from a prepared statement, which may be cached as part of the connection implementation

> Both queries will be bound to the same prepared statement internally, so this will expose the result of the second query instead of the first one

If I understand correctly, the limitation that we cannot execute two queries at the same time on the same connection is not implied by prepared statements and caching, but by the fact that because we're doing SQL, there can not be several queries executed at the same time by a single connection, as the order matters (`BEGIN`, `COMMIT`, `ROLLBACK`, etc...). (I feel like we'd have ways around SQLite prepared statements being tied to query executions otherwise, like prepared statements pools by query type.)

Given that, it looks like having these functions require `&mut Connection` is the natural way to go. It would also (maybe?) remove the need for all these caching `RefCell`s in the connection, so that would be easier to manipulate.
That also seems more correct as it would also naturally prevent stuff like this from compiling:
```rust
thread_local! {
   static ref CONN: Connection = /* ... */;
}
struct T;
impl FromSqlRow<ST, Backend> for T {
    // do stuff with CONN
}
```
which depending on the way the connection is implemented (if e.g. with Postgres we start reading in [single row mode](https://www.postgresql.org/docs/12/libpq-single-row-mode.html) and deserializing on the fly to minimize RAM usage and favor CPU cache usage, or to implement this feature of iterator support) may be incorrect.

However I'm not sure how that would interact with the fact that metadata lookup requires making queries on the connection as well, but that's probably not a blocker, although it may be an important thing to consider in the design.

Regarding backwards-compatibility, I'm never too attached to it but I believe Ctrl+Shift+F would do most of the work (at least for us, probably not for the projects that have been using `.execute(&conn)` everywhere instead of `let conn = &pool.get()?;` though).

**Reply by Ten0** [reply]
Looks like it would require change on these kinds of patterns as well:
https://github.com/diesel-rs/diesel/blob/c22f9fcc2395d2d990b4b180593954315ec8d207/diesel/src/connection/mod.rs#L101-L119
`f` would probably have to take `&mut Connection`.
That would however remove the need to create an intermediate variable when opening transactions using connections that comme from pools, so it would probably often make for a better interface anyway:
```rust
pool.get()?.transaction(|conn| { /* do stuff */});
```
It's also already an abstraction we have in our codebase: `diesel_helpers::with_serializable(|db| { /* do stuff */ })` has always been easier to use for us. Maybe others have similar abstractions as well?

**Reply by weiznich** [reply]
Another thing that needs to change is how we implement the transaction function internally. Currently most of the actual handling is provided by a transaction handler, that internally calls functions on `Connection`, while it is stored inside the same connection. This will not work anymore for obvious reasons as soon as we require mutable/unique references for calling anything on connections. I'm not really sure what's the best way to solve this problem.
I've put together a free standing toy example in the playground [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9d1a262133bbeac0e47290be8b60becc). Feel free to tweak the actual relation between `Connection` and `TransactionManager` there to get a better understanding of the problem. 
That written I would prefer to have some design that provides an abstraction similar to the existing one, so that the actual transaction handling implementation can be shared between different database implementations. I believe transactions are something fundamental to databases, therefore we should get them as right as possible.

**Reply by pksunkara** [reply]
I don't understand why we need to tie those two traits together. I made them a little bit independent and it's working [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=badf14f914c6dd02db90fab0c8577c8f).

**Reply by weiznich** [reply]
[Real implementations](https://github.com/diesel-rs/diesel/blob/master/diesel/src/connection/transaction_manager.rs#L45) are a bit more complicated. That one needs to track how deep transactions are nested for a specific connection.

**Reply by Ten0** [reply]
> would probably also be helpful for the async connection implementation

How would it be helpful for the async connection implementation? (Trying to understand the other beneficial effects we want to keep)

**Reply by Ten0** [reply]
> Another thing that needs to change is how we implement the transaction function internally. Currently most of the actual handling is provided by a transaction handler, that internally calls functions on `Connection`, while it is stored inside the same connection. This will not work anymore for obvious reasons as soon as we require mutable/unique references for calling anything on connections. I'm not really sure what's the best way to solve this problem.

Here's what I came up with to solve this issue:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8260e253f3e632badaee627ea167696e

I'm not sure whether this applies as efficiently to the metadata caching though.

**Reply by weiznich** [reply]
> How would it be helpful for the async connection implementation? (Trying to understand the other beneficial effects we want to keep)

An async `Connection` implementation does very likely also require mutable connection references, as otherwise the futures returned by the corresponding functions are not `Send`, due to the missing `Sync` impl for most/all connection types.

> Here's what I came up with to solve this issue:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8260e253f3e632badaee627ea167696e
> 
> I'm not sure whether this applies as efficiently to the metadata caching though.

That would be a solution 🤔 
Just for the record, we should also have an eye on how third party backends implement their transaction managers:
* [Diesel-oci](https://github.com/GiGainfosystems/diesel-oci/blob/22a5c978913e4487910ab2de5ee008095458e370/src/oracle/connection/transaction.rs#L12)
* [Diesel-firebird](https://github.com/fernandobatels/rsfbclient/blob/fb8baefb43438748571636ee91306250135ef296/rsfbclient-diesel/src/fb/transaction.rs#L12)
(All of them seem to use a similar pattern)

That all written, I'm not that happy to introduce another associated type for the connection trait, as this makes it even harder to potentially provide a simple backend independ `BoxableConnection` someday.

The other solution I've found is to just [use an `Arc`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a67bef36d46190ad896ef8544b66675e) to wrap the transaction manager. While this works I'm not sure on the performance implications (probably not that relevant as you need to call `transaction` quite of to manifest this).

**Reply by Ten0** [reply]
> An async `Connection` implementation does very likely also require mutable connection references, as otherwise the futures returned by the corresponding functions are not `Send`, due to the missing `Sync` impl for most/all connection types.

👍

> That all written, I'm not that happy to introduce another associated type for the connection trait, as this makes it even harder to potentially provide a simple backend independ `BoxableConnection` someday.

It seems like the current `fn transaction_manager(&mut self) -> &dyn TransactionManager<Self>;` (or `fn transaction_manager(&mut self) -> Arc<dyn TransactionManager<Self>>;`) already can't be made into a trait object's method, "because method `transaction_manager` references the `Self` type in its return type".

It looks like when this time comes we'll need do something like this, not only for transaction management but also for all sorts of serializations/deserializations, since generics are not more boxable than associated types, and are used even more extensively:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e5ca3d145e18aad6d977a28276e71322

In the end, it looks like this wouldn't have a significant impact on that topic.

> The other solution I've found is to just [use an `Arc`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a67bef36d46190ad896ef8544b66675e) to wrap the transaction manager. While this works I'm not sure on the performance implications (probably not that relevant as you need to call `transaction` quite of to manifest this).

I'd favor the other solution given the following upsides it has over the `Arc`:
 - Better guarantees of consistency regarding the fact that the data used is always the one that's inside the connection
 - Does not introduce cross-thread synchronization complexity (with the `Arc` you'd still have to handle how you update the arbitrary properties that are inside, while with the other solution you directly get an `&mut TransactionManagerData`: https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=6db5058f4d9ee10e4235fecde6a1725d). How should that be done? With a mutex? probably hurts async as well...
 - We don't want the connection to be sent to a separate thread while we still own the transaction manager, and it doesn't make sense to have a transaction manager allocated somewhere while the connection was dropped.
 - Zero dynamic dispatch, naturally `Send`/`Sync` and correct lifetime, no `Cell`/`Mutex`/`AtomicXXX` => very simple to use, both when using and when implementing a transaction manger

> we should also have an eye on how third party backends implement their transaction managers

Looking at these links it looks like these could indeed easily be adapted to the proposed framework, even getting rid of the `self.transaction_depth.set(self.transaction_depth.get() + by)` patterns.

Note: Rust playground seems to have a hard time ATM but you can probably run locally those that don't run

**Reply by weiznich** [reply]
> It seems like the current fn transaction_manager(&mut self) -> &dyn TransactionManager<Self>; (or fn transaction_manager(&mut self) -> Arc<dyn TransactionManager<Self>>;) already can't be made into a trait object's method, "because method transaction_manager references the Self type in its return type".
> 
> It looks like when this time comes we'll need do something like this, not only for transaction management but also for all sorts of serializations/deserializations, since generics are not more boxable than associated types, and are used even more extensively:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e5ca3d145e18aad6d977a28276e71322

Currently the main issue here is that this does not compose well with `BoxableConnection` introduced for the `Migration` rewrite. That type cannot do much yet, but allows at least the access to a full connection type instead of the old `SimpleConnection` hack. Having another associated type on `Connection` will make this pattern likely harder. (Not sure, need to experiment with that a bit more. Maybe we could just sidestep this issue by not making `Connection` a supertrait of `BoxableConnection` 🤔 )
Other than that I believe at least in theory it is possible to implement `Connection` for `dyn BoxableConnection` itself (which opens the question how to implement the methods, but that's a technical problem, not a type level thing). Also passing a not `Sized` self a type argument to a trait should be fine, as long as the trait accepts explicitly `?Sized` types there. (That's not true for `TransactionManager` yet, but I don't see why this shouldn't be possible in principle).

As unrelated comment: If we go down the route to provide a generic boxable connection type. I would rather not get rid of `type Backend`, but introduce a `GenericBackend` variant and do runtime dispatch based on that type. I should note that this is definitively a unfinished idea yet, so this should not stop us to do necessary changes now.

That all written: The `Arc` approach has definitively a lot problems and seems like overkill for me, which means I'm likely in favor with approach based on your idea (maybe by separating the associated constant into a separate trait or something like this). If someone is interested in playing a bit around with some proof of concept for making `Connection` use `&mut self` references that would be welcome.

**Reply by weiznich** [reply]
I've opened https://github.com/diesel-rs/diesel/pull/2762 to make the `Connection` implementations take a mutable reference. If that PR is merged we can do the iterator stuff in a followup PR.

**Reply by hasezoey** [reply]
i assume this is now "fixed" / "implemented" since `load_iter` is a thing now?

**Reply by weiznich** [reply]
Yes

**Reply by Ten0** [reply]
Although with regards to the initial requirement, this may - in particular for the PG backend - still allocate the memory upfront in a libpq specific C struct, only the rust deserialization will be made through an iterator.
